tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/ProtoSeqOps.scala

```diff
@@ -0,0 +1,268 @@
+package at.forsyte.apalache.tla.bmcmt.rules.aux
+
+import at.forsyte.apalache.tla.bmcmt.types.{CellT, UnknownT}
+import at.forsyte.apalache.tla.bmcmt.{Arena, ArenaCell, SymbState, SymbStateRewriter}
+import at.forsyte.apalache.tla.lir.TypedPredefs._
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir._
+
+/**
+ * Proto sequences that contain the absolute minimum for implementing TLA+ sequences. A proto sequence is essentially an
+ * array of fixed size (which we call 'capacity'). A sequence is a pair that consists of the length (an integer cell)
+ * and a proto sequence (a cell). The important property of the proto sequences is that they appear only in the arenas.
+ * Hence, the complexity of iterative operations over proto sequences does not directly propagate to SMT. All operators
+ * over sequences are expressible as operations over proto sequences, which makes proto sequences a nice layer of
+ * abstraction.
+ *
+ * @param rewriter
+ *   a symbolic rewriter
+ * @author
+ *   Igor Konnov
+ */
+class ProtoSeqOps(rewriter: SymbStateRewriter) {
+
+  /**
+   * Create a proto sequence of the given capacity and initialize its elements with the function `mkElem`.
+   *
+   * @param state
+   *   a symbolic state to start with
+   * @param capacity
+   *   fixed capacity of the proto sequence
+   * @param mkElem
+   *   elements constructor receiving a state and an index-base-1
+   * @return
+   *   a new symbolic state and the cell of the proto sequence
+   */
+  def make(state: SymbState, capacity: Int, mkElem: (SymbState, Int) => (SymbState, ArenaCell)): SymbState = {
+    // The proto sequence is a cell of unknown type, not introduced in SMT.
+    val afterAppend = state.updateArena(_.appendCellNoSmt(UnknownT()))
```

Comment:
It's probably an outdated name. Introduced `nextState` right away.

Reply:
As for the cell of unknown type, yes, it is to avoid introducing one more cell type. I want to keep it different from `Seq(_)`, so if the cell of a proto sequence is accidentally used when a sequence is expected, the code should crash.

Reply:
Why not introduce a proper cell type for it?

Reply:
We have to store the elements in the arena somehow. The old implementation did this by keeping `start :: end :: elems` and it was always a bit fragile to work with this list. I find it nicer to have a separate cell, which is not mapped to SMT anyways.